src/frontend.c

```diff
@@ -82,7 +90,210 @@ char *logfile_path = "out.txt";
 FILE *logfile = NULL;
 #endif
 
+// if version isn't defined by the Makefile
+#ifndef VERSION
+#define VERSION "unknown"
+#endif
+
+// cli pieces
+const char *program_name = "collascii";
+const char *program_version = VERSION;
+const char *program_bug_address = "<https://github.com/olin/collascii/issues/>";
+const char *program_description = "a collaborative ascii editor";
+const char *program_doc =
+    "COLLASCII\n\n"
+    "\"The Future Editor of Yesterday, Tomorrow!\"\n\n"
+    "By Matthew Beaudouin-Lafon, Evan New-Schmidt, and Adam Novotny\n\n"
+    "A brief introduction to the interface:\n"
+    "- arrow keys to move the cursor\n"
+    "- enter a character by typing it, unless the mode specifies otherwise\n"
+    "- <TAB> to switch between modes\n"
+    "- <CTRL-C> to quit\n"
+    "- <CTRL-R> to read from the file\n"
+    "- <CTRL-S> to write to the file\n"
+    "- <PGUP>/<PGDOWN> move up/down a screen height\n"
+    "- <SHIFT-LEFT>/<SHIFT-RIGHT> move left/right a screen width\n";
+
+// representation of desired state from the cmdline interface
+// see the creatively-named "arguments" struct in `main` for the defaults
+typedef struct {
+  int height, width;    // dimensions of initial canvas (use `DEFAULT_` if 0)
+  int x, y;             // initial cursor position
+  bool load_file;       // read from file at start
+  bool use_stdin;       // read from stdin instead of file_name
+  char *filename;       // file to save to/read from
+  bool connect_remote;  // connect to a remote server
+  char *remote_host;
+  char *remote_port;
+} arguments_t;
+
+/* Make sure the arguments struct is valid and in a non-conflicting state.
+ *
+ * Return values are meant for an argp parser function.
+ *
+ * TODO: fill this out
+ */
+int validate_arguments(arguments_t *arguments) { return 0; }
+
+void parse_args(int argc, char *argv[], arguments_t *arguments) {
+  struct arg_lit *help, *version, *usage;
+  struct arg_int *width, *height;
+  struct arg_file *file;
+  struct arg_str *server, *port;
+  struct arg_end *end;
+
+  void *argtable[] = {
+      help = arg_litn(NULL, "help", 0, 1, "display this help and exit"),
+      usage =
+          arg_litn(NULL, "usage", 0, 1, "display brief usage info and exit"),
+      version =
+          arg_litn(NULL, "version", 0, 1, "display version info and exit"),
+      width = arg_intn("w", "width", "<n>", 0, 1, "initial width of canvas"),
```

Comment:
Is it worth saying that 0 width goes to default value?

Reply:
I'm inclined to leave out the 0 bit, include a message about the default canvas size, and reword the width and height args to say they resize/set a custom value.